wvd-templates/DSC/Functions.ps1

```diff
@@ -222,13 +222,159 @@ function ExtractDeploymentAgentZipFile {
+    }
+    else {
+        $PSBoundParameters.Remove('ServicePrincipal')
+        $PSBoundParameters.Remove('TenantId')
+        Write-Log -Message "Authenticating using user $($Credential.username)"
+    }
+
+    $authentication = $null
+    try {
+        $authentication = Add-RdsAccount @PSBoundParameters
+        if (!$authentication) {
+            throw $authentication
+        }
+    }
+    catch {
+        $errMsg = "Windows Virtual Desktop Authentication Failed, Error:`n$($_ | Out-String)"
+        Write-Log -Error "$errMsg"
+        throw "$errMsg"
+    }
+    Write-Log -Message "Windows Virtual Desktop Authentication successfully Done. Result:`n$($authentication | Out-String)"
+}
+
+function SetTenantContextAndValidate {
+    param(
+        [Parameter(mandatory = $true)]
+        [string]$definedTenantGroupName,
+
+        [Parameter(mandatory = $true)]
+        [string]$TenantName
+    )
+    #//todo refactor
+    #//todo try catch ?
+    # Set context to the appropriate tenant group
+    $currentTenantGroupName = (Get-RdsContext).TenantGroupName
+    if ($definedTenantGroupName -ne $currentTenantGroupName) {
+        Write-Log -Message "Running switching to the $definedTenantGroupName context"
+        Set-RdsContext -TenantGroupName $definedTenantGroupName
+    }
+    try {
+        $tenants = Get-RdsTenant -Name "$TenantName"
+        if (!$tenants) {
+            Write-Log "No tenants exist or you do not have proper access."
+            #//todo throw ?
+        }
+    }
+    catch {
+        #//todo refactor msg ?
+        Write-Log -Message $_
+        throw $_
+    }
+}
+
+function ImportRDPSMod {
+    param(
+        [string]$Source = 'attached',
+        [string]$ArtifactsPath
+
+        # [Parameter(mandatory = $false)]
+        # [string]$DeployAgentLocation = 'C:\DeployAgent'
+    )
+
+    #//todo move outside
+    # Write-Log -Message "Creating a folder inside rdsh vm for extracting deployagent zip file"
+    # ExtractDeploymentAgentZipFile -ScriptPath $ScriptPath -DeployAgentLocation $DeployAgentLocation
+    
+    # Write-Log -Message "Changing current folder to Deployagent folder: $DeployAgentLocation"
+    # Set-Location "$DeployAgentLocation"
+    
+    if ($Source -eq 'attached') {
+        if ((-not $ArtifactsPath) -or (-not (test-path $ArtifactsPath))) {
+            throw "invalid param: ArtifactsPath = '$ArtifactsPath'"
+        }
+
+        # Locating and extracting PowerShellModules.zip
+        $ZipName = 'PowerShellModules.zip'
+        Write-Log -Message "Locating '$ZipName' within Custom Script Extension folder structure: '$ArtifactsPath'"
+        $ZipPath = (Get-ChildItem "$ArtifactsPath\" -Filter $ZipName -Recurse | Select-Object -First 1).FullName
+        if ((-not $ZipPath) -or (-not (Test-Path $ZipPath))) {
+            throw "'$ZipName' file not found at '$ArtifactsPath'"
+        }
+        
+        $Path = 'C:\PowerShellModules\'

```

Comment:
Instead of this hardcoded path maybe use a folder under $ArtifactsPath, this has the advantages of:
It helps keep all of the files created by the ARM template in the same folder which allows for easier clean up if needed.
If C:\PowerShellModules\ already exists on the image that is being deployed this wold remove it.
If the script doesn't have permission to create folders on C drive this would fail.

I know that this code already creates C:\DeployAgent but we should eventually move that to be under a folder relative to where the ARM template code is extracted to.

Reply:
I agree with your suggestion but the original scripts also extract to C:\DeployAgent. I guess there might be some reason why it was done that way. So I didn't want to refactor anything that was already working at least for this PR. Moving everything to working directory should be part of future improvements (diff PR).

Reply:
Sounds good but I think for now C:\PowerShellModules\ should be renamed to something less likely to already exist like C:\WVDTemplatePowerShellModules\